src/Features/CSharp/Portable/UsePatternCombinators/CSharpUsePatternCombinatorsAnalyzer.cs

```diff
@@ -0,0 +1,203 @@
+            };
+        }
+
+        private AnalyzedPattern? ParseRelationalPattern(IBinaryOperation op)
+        {
+            return DetermineConstant(op) switch
+            {
+                ConstantResult.Left when op.LeftOperand.Syntax is ExpressionSyntax left
+                    => new Relational(Flip(op.OperatorKind), left),
+                ConstantResult.Right when op.RightOperand.Syntax is ExpressionSyntax right
+                    => new Relational(op.OperatorKind, right),
+                _ => null
+            };
+        }
+
+        private AnalyzedPattern? ParseConstantPattern(IBinaryOperation op)
+        {
+            return DetermineConstant(op) switch
+            {
+                ConstantResult.Left when op.LeftOperand.Syntax is ExpressionSyntax left
+                    => new Constant(left),
+                ConstantResult.Right when op.RightOperand.Syntax is ExpressionSyntax right
+                    => new Constant(right),
+                _ => null
+            };
+        }
+
+        private static bool IsRelationalOperator(BinaryOperatorKind operatorKind)
+        {
+            switch (operatorKind)
+            {
+                case LessThan:
+                case LessThanOrEqual:
+                case GreaterThanOrEqual:
+                case GreaterThan:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Changes the direction the operator is pointing
+        /// </summary>
+        public static BinaryOperatorKind Flip(BinaryOperatorKind operatorKind)
+        {
+            return operatorKind switch
+            {
+                LessThan => GreaterThan,
+                LessThanOrEqual => GreaterThanOrEqual,
+                GreaterThanOrEqual => LessThanOrEqual,
+                GreaterThan => LessThan,
+                var v => throw ExceptionUtilities.UnexpectedValue(v)
+            };
+        }
+
+        private static bool IsConstant(IOperation operation)
+        {
+            // Constants do not propagate to conversions
+            return operation is IConversionOperation op
+                ? IsConstant(op.Operand)
+                : operation.ConstantValue.HasValue;
+        }
+
+        private bool CheckTargetExpression(IOperation operation)
+        {
+            if (operation is IConversionOperation { IsImplicit: false } op)
+            {
+                // Unwrap explicit casts because the pattern will emit those anyways
+                operation = op.Operand;
+            }
+
+            if (!(operation.Syntax is ExpressionSyntax expression))
+            {
+                return false;
+            }
+
+            // If we have not figured the target expression yet,
+            // we will assume that the first expression is the one.
+            if (_targetExpression is null)
+            {
+                _targetExpression = expression;
```

Comment:
fwiw, i didn't like this in the existing feature.  having the analysis mutate as it goes along makes it very finicky wrt changing the order of callers.  I think i'd prefer a pass which first tries to find the target.  then a stateless pass which rewrites given that it knows what the target is.